plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/DeployPropertyPage.java

```diff
@@ -105,69 +110,30 @@ public void run() {
 
   @Override
   public boolean performOk() {
-    // Content can be null if Properties Page is opened and App Engine facets are uninstalled
     if (isValid()) {
-      if (content == null) {
-        return true;
-      } else {
-        return content.savePreferences();
-      }
+      return getActivePanel().savePreferences();
     }
     return false;
   }
 
   @Override
   protected void performDefaults() {
-    content.resetToDefaults();
+    // should we reset all panels?
+    getActivePanel().resetToDefaults();
     super.performDefaults();
   }
 
   @Override
   public void dispose() {
-    if (content != null) {
-      // Content can be null if Properties Page is opened and App Engine facets are uninstalled
-      content.dispose();
-    }
+    blankPreferencesPanel.dispose();
+    standardPreferencesPanel.dispose();
+    flexPreferencesPanel.dispose();
+    databindingSupport.dispose();
     super.dispose();
   }
 
-  @Override
-  public void setErrorMessage(String newMessage) {
-    if (!canSetMessage) {
-      return;
-    }
-    super.setErrorMessage(newMessage);
-  }
-
-  @Override
-  public void setMessage(String newMessage, int newType) {
-    if (!canSetMessage) {
-      return;
-    }
-    super.setMessage(newMessage, newType);
-  }
-
-  private DeployPreferencesPanel getPreferencesPanel(IProject project,
-      IFacetedProject facetedProject, Composite container) {
-
-    IGoogleLoginService loginService =
-        PlatformUI.getWorkbench().getService(IGoogleLoginService.class);
-    IGoogleApiFactory googleApiFactory =
-        PlatformUI.getWorkbench().getService(IGoogleApiFactory.class);
-
-    if (AppEngineStandardFacet.hasAppEngineFacet(facetedProject)) {
-      setTitle(Messages.getString("standard.page.title"));
-      return new StandardDeployPreferencesPanel(
-          container, project, loginService, getLayoutChangedHandler(), false /* requireValues */,
-          new ProjectRepository(googleApiFactory));
-    } else if (AppEngineFlexFacet.hasAppEngineFacet(facetedProject)) {
-      setTitle(Messages.getString("flex.page.title"));
-      return new FlexDeployPreferencesPanel(container, project);
-    } else {
-      logger.log(Level.WARNING,
-          "App Engine Deployment property page is only visible if project contains an App Engine facet");
-      return null;
-    }
+  private DeployPreferencesPanel getActivePanel() {
+    return (DeployPreferencesPanel) stackLayout.topControl;
   }
 
   @Override
```

Comment:
do we need to update the panels when setVisible(false)? I am thinking that when setVisible(false) is called, it is either the panel was already invisible (which we don't care too much about) or it was inimitably visible, which at that point it would have the correct configuration. And then the next time setVisible(true) is called, the configuration will be updated if necessary.